File: julia/2015/day06a/day06a.jl
#!/usr/bin/env julia

struct Position
  x :: Int32
  y :: Int32
end

struct Bounds
  upperLeft :: Position
  lowerRight :: Position
end

function usage()
    progname = PROGRAM_FILE
    println(stderr, "usage: $progname <input file>")
    exit(1)
end

function perform(grid::BitMatrix, action::String, bounds::Bounds)
    for row in bounds.upperLeft.y:bounds.lowerRight.y
        for col in bounds.upperLeft.x:bounds.lowerRight.x
            if action == "turn on"
              grid[row, col] = true
            elseif action == "turn off"
              grid[row, col] = false
            elseif action == "toggle"
              grid[row, col] = !grid[row, col]
            end
        end
    end
end

function process(content::String)::Int32
    grid = falses(1000, 1000)
    for line in split(content, '\n', keepempty=false)
        matches = match(r"(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)", line)
        if matches != nothing
            action :: String = matches.captures[1]
            bounds = Bounds(Position(parse(Int32, matches.captures[2]) + 1,
                                     parse(Int32, matches.captures[3]) + 1),
                            Position(parse(Int32, matches.captures[4]) + 1,
                                     parse(Int32, matches.captures[5]) + 1))
            perform(grid, action, bounds)
        end
    end
    return count(==(true), grid)
end

function main()
    args = ARGS
    if length(args) < 1
        usage()
    end
    filename = args[1]
    try
        content = read(filename, String)
        result = process(content)
        println("result = $result")
    catch e
        println(stderr, "error reading file: $e")
        exit(1)
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: python/2015/day06a/day06a.py
#!/usr/bin/env python3

import re
import sys

def usage():
    progname = sys.argv[0]
    print(f'usage: {progname} <input file>')
    sys.exit(1)

def perform(grid, action, r1, c1, r2, c2):
    for row in range(r1, r2+1):
        for col in range(c1, c2+1):
            if action == 'turn on':
                grid.add((row, col))
            elif action == 'turn off':
                grid.discard((row, col))
            elif action == 'toggle':
                grid.discard((row, col)) if (row, col) in grid else grid.add((row, col))

def process(filename):
    grid = set()
    pattern = re.compile(r'(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)')
    with open(filename, 'r') as infile:
        for line in infile:
            matches = pattern.match(line)
            if matches != None:
                action = matches[1]
                r1 = int(matches[2])
                c1 = int(matches[3])
                r2 = int(matches[4])
                c2 = int(matches[5])
                perform(grid, action, r1, c1, r2, c2)
    return len(grid)

def main():
    if len(sys.argv) < 2:
        usage()
    filename = sys.argv[1]
    result = process(filename)
    print(f'result = {result}')

if __name__ == '__main__':
    main()
